daemon/containerd/image.go

```diff
@@ -1,11 +1,293 @@
 package containerd
 
 import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"regexp"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/content"
+	cerrdefs "github.com/containerd/containerd/errdefs"
+	containerdimages "github.com/containerd/containerd/images"
+	"github.com/docker/distribution/reference"
+	containertypes "github.com/docker/docker/api/types/container"
 	imagetype "github.com/docker/docker/api/types/image"
+	"github.com/docker/docker/errdefs"
 	"github.com/docker/docker/image"
+	"github.com/docker/docker/layer"
+	"github.com/docker/go-connections/nat"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
 )
 
+var shortID = regexp.MustCompile(`^([a-f0-9]{4,64})$`)
+
 // GetImage returns an image corresponding to the image referred to by refOrID.
-func (i *ImageService) GetImage(refOrID string, options imagetype.GetImageOpts) (retImg *image.Image, retErr error) {
-	panic("not implemented")
+func (i *ImageService) GetImage(ctx context.Context, refOrID string, options imagetype.GetImageOpts) (*image.Image, error) {
```

Comment:
This PR probably depends on the context changes and/or "inspect" change to go in first;

- https://github.com/moby/moby/pull/43939
- https://github.com/moby/moby/pull/43828
- https://github.com/moby/moby/pull/43818


```
#41 0.150 ---> Making bundle: binary (in /build/bundles/binary)
#41 0.186 Building: /build/bundles/binary-daemon/dockerd
#41 0.186 GOOS="" GOARCH="" GOARM=""
#41 44.98 # github.com/docker/docker/daemon/images
#41 44.98 daemon/images/cache.go:19:31: not enough arguments in call to i.GetImage
#41 44.98 	have (string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 	want (context.Context, string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 daemon/images/image_builder.go:171:31: not enough arguments in call to i.GetImage
#41 44.98 	have (string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 	want (context.Context, string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 daemon/images/image_builder.go:215:35: not enough arguments in call to i.GetImage
#41 44.98 	have (string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 	want (context.Context, string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 daemon/images/image_delete.go:67:35: not enough arguments in call to i.GetImage
#41 44.98 	have (string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 	want (context.Context, string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 daemon/images/image_events.go:15:34: not enough arguments in call to i.GetImage
#41 44.98 	have (string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 	want (context.Context, string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 daemon/images/image_history.go:16:31: not enough arguments in call to i.GetImage
#41 44.98 	have (string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 	want (context.Context, string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 daemon/images/image_history.go:72:42: not enough arguments in call to i.GetImage
#41 44.98 	have (string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 	want (context.Context, string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 daemon/images/image_list.go:53:41: not enough arguments in call to i.GetImage
#41 44.98 	have (string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 	want (context.Context, string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 daemon/images/image_list.go:61:40: not enough arguments in call to i.GetImage
#41 44.98 	have (string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 	want (context.Context, string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 daemon/images/image_pull.go:67:33: not enough arguments in call to i.GetImage
#41 44.98 	have (string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 	want (context.Context, string, "github.com/docker/docker/api/types/image".GetImageOpts)
#41 44.98 daemon/images/image_pull.go:67:33: too many errors
#41 ERROR: process "/bin/sh -c hack/make.sh binary" did not complete successfully: exit code: 2
```

Reply:
yes sorry, this was just an initial experiment based on cherry-picking a single commit, shouldhave ben opened as Draft